src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -282,6 +292,139 @@ internal static void RunInternal<TState>(ExecutionContext executionContext, Cont
+                executionContext = null;
+            }
+            else if (executionContext != null)
+            {
+                // Non-Default context to restore
+                currentThread.ExecutionContext = executionContext;
+                if (executionContext.HasChangeNotifications)
+                {
+                    // There are change notifications; trigger any affected
+                    OnValuesChanged(previousExecutionCtx: null, executionContext);
+                }
+            }
+
+            ExceptionDispatchInfo edi = null;
+            try
+            {
+                callback.Invoke(state);
+            }
+            catch (Exception ex)
+            {
+                // Note: we have a "catch" rather than a "finally" because we want
+                // to stop the first pass of EH here.  That way we can restore the previous
+                // context before any of our callers' EH filters run.
+                edi = ExceptionDispatchInfo.Capture(ex);
+            }
+
+            // Enregister currentThread to alternative varibale as crossing EH will have forced it to stack
+            Thread currentThread1 = currentThread;
+            // The common case is that these have not changed, so avoid the cost of a write barrier if not needed.
+            if (currentThread1.SynchronizationContext != null)
+            {
+                // Restore changed SynchronizationContext back to previous
+                currentThread1.SynchronizationContext = null;
+            }
+
+            ExecutionContext currentExecutionCtx = currentThread1.ExecutionContext;
+            if (currentExecutionCtx != null)
+            {
+                // Restore changed ExecutionContext back to Default (null)
+                currentThread1.ExecutionContext = null;
+                if (currentExecutionCtx.HasChangeNotifications)
+                {
+                    // There are change notifications; trigger any affected
+                    OnValuesChanged(currentExecutionCtx, nextExecutionCtx: null);
+                }
+            }
+
+            // If exception was thrown by callback, rethrow it now original contexts are restored
+            edi?.Throw();
+        }
+
+        internal static void RunForThreadPoolUserWorkItem<TState>(ExecutionContext executionContext, Action<TState> callback, in TState state)
+        {
+            Debug.Assert(Thread.CurrentThread.IsThreadPoolThread);
+            Debug.Assert(Thread.CurrentThread.ExecutionContext == null, "ThreadPool thread not on Default ExecutionContext.");
+            Debug.Assert(Thread.CurrentThread.SynchronizationContext == null, "ThreadPool thread not on Default SynchronizationContext.");
+
+            // Capture current thread to local rather than looking it up multiple times
+            Thread currentThread = Thread.CurrentThread;
+            if (executionContext != null && !executionContext.m_isDefault)
```

Comment:
This should never be false; so could just change to Assert

Reply:
If changed to assert, it could probably use a name change as well to reflect that a default context cannot be passed in